### ISSUE TITLE
Design of identType

### ISSUE THREAD

**Issue by anton-trunov** [volodeyka/event-struct#21]
So far we have an idea that `identType` is a type `T` with a function `fresh : seq T -> T` with the property `(fresh xs) \notin xs`.

This, however, presents an issue: the `fresh` function needs to traverse its input list to generate a fresh value but if we need to call it repeatedly to generate a sequence of fresh identifiers, this resulting algorithm is going to be quadratic.

Instead, we could change the type of `fresh` to `T -> T`, i.e. we can generate a new fresh ident given the previous one. (So, for example, an instance of `identType` for `nat` would have `succn` as its `fresh` function.)

Let me outline several property statements that seem to be useful.

1. One of the most general ones (it only requires `T` to be `eqType`) could look like this (`traject` lives in Mathcomp's `path` module):
   ```coq
   forall (n : nat) (x : T), unique (traject fresh x n)
   ```
  Basically, this says that all powers of `fresh` function produce distinct elements.
2. Another option (this would require `T` to be `ordType`) is to say
   ```coq
   forall x : T, x < fresh x
   ```
3. And yet another option would be to make `fresh` of type `nat -> seq T` generating an n-tuple of unique identifiers:
   - `forall n, size (fresh n) = n`;
   - `forall n, unique (fresh n)`.

**Comment by eupp** [comment]
A minor stylistic suggestion.
Can we agree to name the variables of type `identType` as `id`, `id'`, `id1`, `id2` etc, instead of `ident` and its variations?
The `id` is shorter and still has clear semantics. 
If we will choose this convention, I would propose to rename `ident0` into `id0`. 

Other options for `ident0` (if we don't want to have char `0` in the name) are `first` or `start`.

@anton-trunov @volodeyka @dmitromikh

**Comment by volodeyka** [comment]
Should we close it for now?

**Comment by volodeyka** [comment]
> A minor stylistic suggestion.
> Can we agree to name the variables of type `identType` as `id`, `id'`, `id1`, `id2` etc, instead of `ident` and its variations?
> The `id` is shorter and still has clear semantics.

`id` conflicts with identity function, unfortunately. And in most cases, we are looking at elements of `identType` as on events, and name them `e`, `e1`, `e2`... So maybe `ident` is better. 
But if we want to somehow name a fresh identifier we can use something like `fresh_id` (in `transitionsystem.v` for example)

**Comment by eupp** [comment]
> `id` conflicts with identity function, unfortunately. 

Fair enough. Let's stick with `ident` and its derivatives.

> But if we want to somehow name a fresh identifier we can use something like `fresh_id` (in `transitionsystem.v` for example)

**Comment by volodeyka** [comment]
> In that context, `fresh_id` looks to verbose, comparing to other event identifiers like `e`, `e1`, `e2`, etc. So we need to invent a better convention. We can just name the fresh event simply as `e` (not so good), `en` (`n` here to reflect it is `n`-th event added on the `n`-th step), or `e_fr` (short for `e_fresh`). I personally like `en`. Other suggestions are welcome.

I vote for `e_fr` or `efr`

**Comment by anton-trunov** [comment]
In that context we never use `fresh`. How about using it? It's just one symbol longer than `e_fr` but actually readable.

**Comment by eupp** [comment]
> > In that context we never use `fresh`. How about using it? It's just one symbol longer than `e_fr` but actually readable.
> 
> It was my first solution :)

Did you proposed it in some PR?
Sorry if I missed that and then have repeated it here. 

It looks like we still haven't reached consensus.
I vote for `en`, @volodeyka for `e_fr`/`efr`, @dmitromikh is silent.
Thus @anton-trunov has casting vote :)

**Comment by anton-trunov** [comment]
I'd rather have `fresh` spelled out fully (as most papers seem to do), but if need be then something that has at least `fr` in it should be acceptable. Let me re*fresh* the context of `fresh_id` usage and I'll get back to to this issue.

**Comment by eupp** [comment]
> I'd rather have `fresh` spelled out fully (as most papers seem to do), but if need be then something that has at least `fr` in it should be acceptable. Let me re_fresh_ the context of `fresh_id` usage and I'll get back to to this issue.

Ok, thanks!

Just in case: we speak about the particular context, inside the construction of event structures. 
There, we primarly use `e` to denote events, and thus `fresh_id` stands out. 
An example of the problem https://github.com/volodeyka/event-struct/blob/294771c1e504d73e05a4ab29ad8a966db5d57ae5/transitionsystem.v#L176

In general I'm ok with `fresh_id`, it's just this particular use case that bothers me.

**Comment by volodeyka** [comment]
> I vote for `en`, @volodeyka for `e_fr`/`efr`, @dmitromikh is silent.

I vote for `fresh`

**Comment by eupp** [comment]
With @volodeyka we have recently faced a new challenge related to `identType`. 
Let `E : identType`, `f` and `g` are two finitely supported functions `f g : { fsfun E -> D }`, an assume that the support of `f` is less than that of `g`, i.e. `supp f <= supp g`. 
The let `h : supp f -> E` be a function that is injective on the support of `f`. We want to have a canonical way to extend `h` to the whole type `E` so that the resulting function is still injective. 
An intuitive way to do that is to map all elements outside of the support of `f` to elements outside of the support of `g`. 
However, it is not possible to do so with the current design. 

Alternatively, assume `E == nat`. Then there is an easy way to do what we want. Let `f` be defined on `[0 ... n]` and let `g` be defined on `[0 .. n + k]`. Then we can map `n+i` to `n + i + k`. 

In order to abstract the described approach, we need to require more structure on `E : identType`. 
A new proposal is to require that `identType` be a countable infinite type. 
Countability would give us `pickle : E -> nat` and `unpickle : nat -> E`. 
Moreover, we will get `ident0` and `fresh` for free:

- `ident0 := unpickle 0`
- `fresh x := unpickle (1 + pickle x)`

 Or else, if `pickle` and `unpickle` are computationally intensive functions, we can just require that 
 `fresh x = unpickle (1 + pickle x)`.

This new encoding of `identType` seems to be compatible with 2 most important instances: `nat`  and `seq A` (i.e. strings over an alphabet of `A`). 

cc @anton-trunov @volodeyka

**Comment by eupp** [comment]
Another issue related to `identType`. 
We currently require `identType` to be instance of `porderType`. 
This gives as order notations: `x <= y`. 
However, in `porf_eventstruct.v` we have an event structure over events of type `E : identType`. 
We then declare another instance of `porderType` over events where partial order is causality order. 
Currently, it enforces us to use `disp` machinery and re-declare new notations to distinguish two partial order. 
Together with @dmitromikh for now we decided to use `e1 <=^c e2` notation for causality. 

We expect the causality order to be use more often than the order induced by `identType`. 
Thus, it would be nice in future to swap the two, i.e to have notation `e1 <= e2` for causality order and notation `e1 <=^i e2` for the partial order on identifiers.
